apps/emqx_gcp_device/src/emqx_gcp_device_authn.erl

```diff
@@ -0,0 +1,233 @@
+                        reason => "not a JWT",
+                        client => ClientInfo
+                    }),
+                    ignore;
+                {false, expired} ->
+                    ?tp(authn_gcp_device_check, #{
+                        result => not_authorized, reason => "expired JWT", client => ClientInfo
+                    }),
+                    ?TRACE_AUTHN_PROVIDER(info, "auth_failed", #{
+                        reason => "expired JWT",
+                        client => ClientInfo
+                    }),
+                    {error, not_authorized}
+            end;
+        not_a_gcp_clientid ->
+            ?tp(authn_gcp_device_check, #{
+                result => ignore, reason => "not a GCP ClientId", client => ClientInfo
+            }),
+            ?TRACE_AUTHN_PROVIDER(debug, "auth_ignored", #{
+                reason => "not a GCP ClientId",
+                client => ClientInfo
+            }),
+            ignore
+    end.
+
+check_jwt(ClientInfo, DeviceId) ->
+    case emqx_gcp_device:get_device_actual_keys(DeviceId) of
+        not_found ->
+            ?tp(authn_gcp_device_check, #{
+                result => ignore, reason => "key not found", client => ClientInfo
+            }),
+            ?TRACE_AUTHN_PROVIDER(debug, "auth_ignored", #{
+                reason => "key not found",
+                client => ClientInfo
+            }),
+            ignore;
+        Keys ->
+            case any_key_matches(Keys, ClientInfo) of
+                true ->
+                    ?tp(authn_gcp_device_check, #{
+                        result => ok, reason => "auth success", client => ClientInfo
+                    }),
+                    ?TRACE_AUTHN_PROVIDER(debug, "auth_success", #{
+                        reason => "auth success",
+                        client => ClientInfo
+                    }),
+                    ok;
+                false ->
+                    ?tp(authn_gcp_device_check, #{
+                        result => {error, bad_username_or_password},
+                        reason => "no matching or valid keys",
+                        client => ClientInfo
+                    }),
+                    ?TRACE_AUTHN_PROVIDER(info, "auth_failed", #{
+                        reason => "no matching or valid keys",
+                        client => ClientInfo
+                    }),
+                    {error, bad_username_or_password}
+            end
+    end.
+
+any_key_matches(Keys, ClientInfo) ->
+    lists:any(fun(Key) -> key_matches(Key, ClientInfo) end, Keys).
+
+key_matches(KeyRaw, #{password := Jwt} = _ClientInfo) ->
+    Jwk = jose_jwk:from_pem(KeyRaw),
+    case jose_jws:verify(Jwk, Jwt) of
+        {true, _, _} ->
+            true;
+        {false, _, _} ->
+            false
+    end.
+
+gcp_deviceid_from_clientid(#{clientid := <<"projects/", RestClientId/binary>>}) ->
+    case binary:split(RestClientId, <<"/">>, [global]) of
```

Comment:
nit: could use the `emqx_topic:tokens/1`

Reply:
It could be used, but, as `RestClientId` isn't a topic, I think it would be weird to use `emqx_topic:tokens` here. Also, if in the future we add a new separator for topic, it might have some non expected impact here (ok, it probably will not happen, but just to mention it 😅)